biahub/track.py

```diff
@@ -0,0 +1,694 @@
+    if input_channels['virtual_stain'] is None:
+        raise ValueError("Virtual Staining input channels is required.")
+    # Check if the tracking channel is present in the virtual staining channels
+    if input_channels['tracking'] is None:
+        raise ValueError("Tracking input channels is required.")
+    elif len(input_channels["tracking"]) != 1:
+        raise ValueError("Only one channel is allowed and required for tracking.")
+    elif input_channels["tracking"][0] not in input_channels["virtual_stain"]:
+        raise ValueError("Tracking channel not found in Virtual Stain input channels.")
+
+    click.echo(f"Loading data from: {input_vs_path}...")
+    vs_dataset = open_ome_zarr(input_vs_path)
+    T, C, Z, Y, X = vs_dataset.data.shape
+    channel_names = vs_dataset.channel_names
+    click.echo(f"Virtual Stanining Channel names: {channel_names}")
+    yx_scale = vs_dataset.scale[-2:]
+
+    # Define output metadata
+    output_channels = []
+    for channel in input_channels["tracking"]:
+        output_channels.append(f"{channel}_labels")
+
+    output_metadata = {
+        "shape": (T, 1, 1, Y, X),
+        "chunks": None,
+        "scale": vs_dataset.scale,
+        "channel_names": output_channels,
+        "dtype": np.uint32,
+    }
+
+    create_empty_hcs_zarr(
+        store_path=output_dirpath, position_keys=[position_key], **output_metadata
+    )
+
+    # Load virtual staining data
+    for channel in input_channels["virtual_stain"]:
+        data_dict[channel] = vs_dataset[0][:, channel_names.index(channel), z_slices]
+
+    # Apply virtual staining projection function
+    if vs_projection_function is not None:
+        click.echo(
+            f"Applying {vs_projection_function.function} projection to the virtual staining data..."
+        )
+
+        projection = getattr(np, vs_projection_function.function)  # Convert string to function
+        kwargs = vs_projection_function.kwargs if vs_projection_function.kwargs else {}
+
+        for input_array in vs_projection_function.input_arrays:
+            data_dict[input_array] = projection(data_dict[input_array], **kwargs)
+
+    # Preprocess to get the the foreground and multi-level contours
+    click.echo("Preprocessing...")
+    if input_segmentation_dirpaths is not None:
+        click.echo(f"Loading segmentation from: {input_segmentation_dirpaths}...")
+        input_segmentation_path = input_segmentation_dirpaths / Path(*position_key)
+
+        label_dataset = open_ome_zarr(input_segmentation_path)
+        label_arr = label_dataset[0][:, 0, :, :, :]  # Shape (T, Z, Y, X)
+
+        # Cast to uint32 in case Cellpose saved as float32
+        label_arr = label_arr.astype(np.uint32)
+
+        click.echo("Converting labels to foreground and contour gradient map...")
+        foreground_mask, contour_gradient_map = labels_to_edges(label_arr, sigma=90)
+
+    else:
+
+        foreground_mask, contour_gradient_map = data_preprocessing(
```

Comment:
Maybe a question for @JoOkuma: Is it better to always do intensity-based segmentation even when cellpose masks are available to get more hypotheses?